design_docs/rsyslog_new_input_output_roles_proposal.md

```diff
@@ -0,0 +1,254 @@
+        target: remote_host_name.remote_domain_name
+        port: 514
+    logs_collections:
+      - name: basic-input0
+        type: basics
+        rsyslog_imjournal_persist_state_interval: 1000
+        rsyslog_imjournal_ratelimit_interval: 10
+        rsyslog_imjournal_ratelimit_burst: 100000
+  - name: file-output
+    type: files
+    logs_collections:
+      - name: basic-input1
+        type: basics
+        rsyslog_imjournal_persist_state_interval: 3000
+        rsyslog_imjournal_ratelimit_interval: 30
+        rsyslog_imjournal_ratelimit_burst: 300000
+```
+
+## To Do's
+
+### Elasticsearch
+There is another output role `elasticsearch`.
+If this proposal is approved,
+I'm planning to rename the 30-output-elasticsearch.conf to 30-output.elasticsearch and add the following rulesets.
+```
+ruleset(name="elasticsearch") {
+  $IncludeConfig /etc/rsyslog.d/*.elasticsearch
+}
+ruleset(name="elasticsearch_files") {
+  $IncludeConfig /etc/rsyslog.d/*.elasticsearch
+  $IncludeConfig /etc/rsyslog.d/*.files
+}
+ruleset(name="elasticsearch_forwards") {
+  $IncludeConfig /etc/rsyslog.d/*.elasticsearch
+  $IncludeConfig /etc/rsyslog.d/*.forwards
+}
+ruleset(name="elasticsearch_files_forwards") {
+  $IncludeConfig /etc/rsyslog.d/*.elasticsearch
+  $IncludeConfig /etc/rsyslog.d/*.files
+  $IncludeConfig /etc/rsyslog.d/*.forwards
+}
+```
+And introduce rsyslog_use_elasticsearch_ruleset parameter to specify the elasticsearch output as follows.
+```
+logs_collections:
+  - name: basic-input
+    type: basics
+    rsyslog_use_files_ruleset: true
+    rsyslog_use_elasticsearch_ruleset: true
+```
+
+Extending this idea, when we add a new output, e.g., kafka to the existing 3 output roles, we are adding following 8 rulesets and deploy the output config file 30-output.kafka.
+```
+ruleset(name="kafka") {
+  $IncludeConfig /etc/rsyslog.d/*.kafka
+}
+ruleset(name="elasticsearch_kafka") {
+  $IncludeConfig /etc/rsyslog.d/*.elasticsearch
+  $IncludeConfig /etc/rsyslog.d/*.kafka
+}
+ruleset(name="files_kafka") {
+  $IncludeConfig /etc/rsyslog.d/*.files
+  $IncludeConfig /etc/rsyslog.d/*.kafka
+}
+ruleset(name="forwards_kafka") {
+  $IncludeConfig /etc/rsyslog.d/*.forwards
+  $IncludeConfig /etc/rsyslog.d/*.kafka
+}
+ruleset(name="elasticsearch_files_kafka") {
+  $IncludeConfig /etc/rsyslog.d/*.elasticsearch
+  $IncludeConfig /etc/rsyslog.d/*.files
+  $IncludeConfig /etc/rsyslog.d/*.kafka
+}
+ruleset(name="elasticsearch_forwards_kafka") {
+  $IncludeConfig /etc/rsyslog.d/*.elasticsearch
+  $IncludeConfig /etc/rsyslog.d/*.forwards
+  $IncludeConfig /etc/rsyslog.d/*.kafka
+}
+ruleset(name="files_forwards_kafka") {
+  $IncludeConfig /etc/rsyslog.d/*.files
+  $IncludeConfig /etc/rsyslog.d/*.forwards
+  $IncludeConfig /etc/rsyslog.d/*.kafka
+}
+ruleset(name="elasticsearch_files_forwards_kafka") {
+  $IncludeConfig /etc/rsyslog.d/*.elasticsearch
+  $IncludeConfig /etc/rsyslog.d/*.files
+  $IncludeConfig /etc/rsyslog.d/*.forwards
+  $IncludeConfig /etc/rsyslog.d/*.kafka
```

Comment:
Does this mean, in the future, if we support 8 different output types, we will need `8!` (8 factorial) different rulesets?  That seems hard to scale and manage . . .
What about `files_elasticsearch`, `forwards_elasticsearch`, etc.?

Reply:
I know...  That's the ugliest part in this proposal...  In terms of the non-alphabetically ordered name, we should be able to sort or map to get the right ruleset name, I'd think.

If we could ignore the non-alphabetically ordered ruleset names, the number would be (1+2+3+...N) = (1+N) * N/2.  If we support 8 output roles, then there would be 36 of them... :(

Reply:
That just seems wrong to me - what if you don't use rulesets so much, or figure out a way to only have a ruleset per output, and combine rulesets with config code e.g. something like
```
if outputs_string contains " kafka " then {
    call kafka
}
if outputs_string contains " elasticsearch " then {
    call elasticsearch
}
```
not sure about the syntax - but the idea is that you don't have inputs strictly bound to rulesets

Reply:
> That just seems wrong to me - what if you don't use rulesets so much, or figure out a way to only have a ruleset per output, and combine rulesets with config code e.g. something like
> 
> ```
> if outputs_string contains " kafka " then {
>     call kafka
> }
> if outputs_string contains " elasticsearch " then {
>     call elasticsearch
> }
> ```

Thanks, @richm!!  My original code was as simple as like this...  Somehow my test failed to work back then, and I thought consecutive calls is not valid.  I tested it again and found it was my test problem... *shame shame* Let me update the code...
```
{% if rsyslog_use_files_ruleset %}
call files
{% endif %}
{% if rsyslog_use_forwards_ruleset %}
call forwards
{% endif %}
```